### PR TITLE
Add Hue/Saturation/Brightness/Contrast adjustment function in HWC for…

### PR DESCRIPTION
**Pull request by ljia5**
… setprop

### CODE REVIEW

File: drm/DrmDevice.h [hunk 1]
```diff
@@ -63,6 +63,26 @@ class DrmDevice {
...
+  auto GetAdjustmentHue() const {
+    return adjustment_hue_;
+  }
+
+  auto GetAdjustmentContrast() const {
+    return adjustment_contrast_;
+  }
+
+  auto GetAdjustmentBrightness() const {
+      return adjustment_brightness_;
```
Comment (feijiang1) [inline drm/DrmDevice.h:83]: space is not aligned

File: hwc2_device/HwcDisplay.cpp [hunk 2]
```diff
@@ -719,6 +723,7 @@ HWC2::Error HwcDisplay::SetColorTransform(const float *matrix, int32_t hint) {
     return HWC2::Error::BadParameter;
 
   color_transform_hint_ = static_cast<android_color_transform_t>(hint);
+
```
Comment (feijiang1) [inline hwc2_device/HwcDisplay.cpp:726]: empty line is not needed

File: drm/DrmDevice.cpp [hunk 3]
```diff
@@ -117,6 +117,77 @@ auto DrmDevice::Init(const char *path) -> int {
...
+  property_get("vendor.hwcomposer.brightness.blue", property, BRIGHTNESS_MAX_NUM_STRING);
+  adjustment_brightness_blue_ = atoi(property) <= 0 ? atoi(BRIGHTNESS_MAX_NUM_STRING) : atoi(property);
+  ALOGD("COLOR_ The property 'vendor.hwcomposer.brightness.blue' value is %d", adjustment_brightness_blue_);
+  adjustment_brightness_red_ &= 0xFF;
+  adjustment_brightness_green_ &= 0xFF;
+  adjustment_brightness_blue_ &= 0xFF;
+  adjustment_brightness_ = (adjustment_brightness_red_  << 16) |
+         (adjustment_brightness_green_ << 8) |
+         (adjustment_brightness_blue_);
+
```
Comment (feijiang1) [inline drm/DrmDevice.cpp:190]: it is better move upper code to a function

File: drm/DrmAtomicStateManager.h [hunk 4]
```diff
@@ -50,6 +51,12 @@ struct AtomicCommitArgs {
   }
 };
 
+struct gamma_colors {
+  float red;
+  float green;
+  float blue;
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.h:57]: class member need add _ as postfix

File: drm/DrmAtomicStateManager.cpp [hunk 5]
```diff
@@ -326,4 +332,347 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+  }
+
+  drmModeObjectSetProperty(pipe_->device->GetFd(), pipe_->crtc->Get()->GetId(), DRM_MODE_OBJECT_CRTC,
+                           pipe_->crtc->Get()->GetGammaLutProperty().id(), lut_blob_id);
+  drmModeDestroyPropertyBlob(pipe_->device->GetFd(), lut_blob_id);
+  return 0;
+}
+
+auto DrmAtomicStateManager::SetColorSaturationHue(void) ->int{
+  double color_transform_matrix_[16] = {1.0, 0.0, 0.0, 0.0,
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:611]: local variable should not have postfix _

File: drm/DrmAtomicStateManager.cpp [hunk 6]
```diff
@@ -326,4 +332,347 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
                                      DRM_MODE_DPMS_ON);
 }
 
+void DrmAtomicStateManager::MatrixMult3x3(const double Matrix1[3][3], const double Matrix2[3][3], double Result[3][3])
+{
+    double Tmp[3][3];
+
+    for (int y = 0; y < 3; y++)
+    {
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:340]: code style is not aligned

File: drm/DrmAtomicStateManager.cpp [hunk 7]
```diff
@@ -196,6 +196,12 @@ auto DrmAtomicStateManager::CommitFrame(AtomicCommitArgs &args) -> int {
     flags |= DRM_MODE_ATOMIC_NONBLOCK;
   }
 
+  if (args.color_adjustment == true)
+  {
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:200]: coding style is not aligned

File: drm/DrmAtomicStateManager.cpp [hunk 8]
```diff
@@ -326,4 +332,347 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+  if (Saturation < 0.75 || Saturation > 1.25)
+    Saturation = 1.0;
+
+  ALOGD("COLOR_ Hue=%f", Hue);
+  ALOGD("COLOR_ Saturation=%f", Saturation);
+
+  if (!ctm) {
+    ALOGE("COLOR_ Cannot allocate CTM memory");
+    return -EINVAL;
+  }
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:419]: such check should be move upper

File: drm/DrmAtomicStateManager.cpp [hunk 9]
```diff
@@ -326,4 +332,347 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
                                      DRM_MODE_DPMS_ON);
 }
 
+void DrmAtomicStateManager::MatrixMult3x3(const double Matrix1[3][3], const double Matrix2[3][3], double Result[3][3])
+{
+    double Tmp[3][3];
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:337]: change Tmp to tmp

**Review by feijiang1 — COMMENTED** [review]
some change

File: drm/DrmAtomicStateManager.cpp [hunk 10]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
                                      DRM_MODE_DPMS_ON);
 }
 
+void DrmAtomicStateManager::MatrixMult3x3(const double matrix_1[3][3], const double matrix_2[3][3], double result[3][3])
+{
+    double tmp[3][3];
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:336]: not aligned

File: drm/DrmAtomicStateManager.cpp [hunk 11]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+    saturation = 1.0;
+  }
+
+  ALOGD("COLOR_ hue=%f", hue);
+  ALOGD("COLOR_ saturation=%f", saturation);
+
+  if (!ctm) {
+    ALOGE("COLOR_ Cannot allocate CTM memory");
+    return -EINVAL;
+  }
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:421]: need remove, already move upper

File: drm/DrmAtomicStateManager.cpp [hunk 12]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+  if (file_brightness != NULL) {
+    read_bytes = fread(buf, 1, 8, file_brightness);
+    if (read_bytes <= 0) {
+      ALOGE("COLOR_ fread saturation error");
+      fclose(file_brightness);
+    }
+    if (atoi(buf) < 0 || atoi(buf) > 255) {
+      brightness_c = 0x80;
+    } else {
+      brightness_c = atoi(buf) &0xFF;
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:646]: & need space

File: drm/DrmAtomicStateManager.cpp [hunk 13]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+  uint32_t brightness_c = drm->GetAdjustmentBrightness();//0x808080;
+
+  file_brightness = fopen("/data/vendor/color/brightness", "r+");
+  if (file_brightness != NULL) {
+    read_bytes = fread(buf, 1, 8, file_brightness);
+    if (read_bytes <= 0) {
+      ALOGE("COLOR_ fread saturation error");
+      fclose(file_brightness);
+    }
+    if (atoi(buf) < 0 || atoi(buf) > 255) {
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:643]: {} only include one line, can remove {}

File: drm/DrmAtomicStateManager.cpp [hunk 14]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+  }
+  free(ctm);
+  return 0;
+}
+
+auto DrmAtomicStateManager::ApplyPendingCTM(
+  struct drm_color_ctm *ctm) -> int {
+  if (pipe_->crtc->Get()->GetCtmProperty().id() == 0) {
+      ALOGE("GetCtmProperty().id() == 0");
+      return -ENOMEM;
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:469]: why return ENOMEM, is there better return value?

File: drm/DrmAtomicStateManager.cpp [hunk 15]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+  }
+
+  uint32_t ctm_id = 0;
+  drmModeCreatePropertyBlob(pipe_->device->GetFd(), ctm, sizeof(drm_color_ctm), &ctm_id);
+  if (ctm_id == 0) {
+    ALOGE("COLOR_ ctm_id == 0");
+    return -ENOMEM;
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:476]: same as upper

File: drm/DrmAtomicStateManager.cpp [hunk 16]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+
+  return 0;
+}
+
+float DrmAtomicStateManager::TransformContrastBrightness(float value, float brightness,
+                                              float contrast) {
+  float result;
+  result = (value - 0.5) * contrast + 0.5 + brightness;
+
+  if (result < 0.0) {
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:491]: {} is not needed

File: drm/DrmAtomicStateManager.cpp [hunk 17]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+
+auto DrmAtomicStateManager::SetColorCorrection(struct gamma_colors gamma,
+                                    uint32_t contrast_c,
+                                    uint32_t brightness_c) ->int{
+  struct drm_color_lut *lut;
+  float brightness[3];
+  float contrast[3];
+  uint8_t temp[3];
+  uint64_t lut_size = 0;
+  int ret =0;
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:522]: =need space

File: drm/DrmAtomicStateManager.cpp [hunk 18]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+  lut = (struct drm_color_lut *)malloc(sizeof(struct drm_color_lut) * lut_size);
+  if (!lut) {
+    ALOGD("Cannot allocate LUT memory");
+    return -ENOMEM;
+  }
+
+  /* Unpack brightness values for each channel */
+  temp[0] = (brightness_c >> 16) & 0xFF;
+  temp[1] = (brightness_c >> 8) & 0xFF;
+  temp[2] = (brightness_c)&0xFF;
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:546]: & need space

**Review by feijiang1 — COMMENTED** [review]
some comments

File: drm/DrmAtomicStateManager.cpp [hunk 19]
```diff
@@ -326,4 +331,345 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+    ALOGD("Cannot allocate ctm memory");
+    return -ENOMEM;
+  }
+
+  file_saturation = fopen("/data/vendor/color/saturation", "r+");
+  if (file_saturation != NULL) {
+    read_bytes = fread(buf, 1, 8, file_saturation);
+    if (read_bytes <= 0) {
+      ALOGE("COLOR_ fread saturation error");
+      fclose(file_saturation);
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:389]: should directly return here

File: drm/DrmAtomicStateManager.cpp [hunk 20]
```diff
@@ -326,4 +331,345 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+    saturation = atof(buf)/100;
+    fclose(file_saturation);
+  }
+
+  file_hue = fopen("/data/vendor/color/hue", "r+");
+  if (file_hue != NULL) {
+    read_bytes = fread(buf, 1, 8, file_hue);
+    if (read_bytes <= 0) {
+      ALOGE("COLOR_ fread hue error");
+      fclose(file_hue);
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:400]: same as upper

**Review by feijiang1 — COMMENTED** [review]
comments

File: drm/DrmAtomicStateManager.cpp [hunk 21]
```diff
@@ -326,4 +331,345 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+  char buf[4096] = {};
+  uint32_t contrast_c = 0x80;
+  uint32_t brightness_c = 0x80;
+
+  file_brightness = fopen("/data/vendor/color/brightness", "r+");
+  if (file_brightness != NULL) {
+    read_bytes = fread(buf, 1, 8, file_brightness);
+    if (read_bytes <= 0) {
+      ALOGE("COLOR_ fread saturation error");
+      fclose(file_brightness);
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:634]: should return here

File: drm/DrmAtomicStateManager.cpp [hunk 22]
```diff
@@ -326,4 +331,345 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+                    (brightness_c));
+    fclose(file_brightness);
+  }
+
+  file_contrast = fopen("/data/vendor/color/contrast", "r+");
+  if (file_contrast != NULL) {
+    read_bytes = fread(buf, 1, 8, file_contrast);
+    if (read_bytes <= 0) {
+      ALOGE("COLOR_ fread hue error");
+      fclose(file_contrast);
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:653]: should return here

File: drm/DrmAtomicStateManager.cpp [hunk 23]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+      tmp[y][x] = matrix_1[y][0] * matrix_2[0][x] + matrix_1[y][1] * matrix_2[1][x] + matrix_1[y][2] * matrix_2[2][x];
+    }
+  }
+
+  memcpy(result, tmp, sizeof(tmp));
+}
+
+void DrmAtomicStateManager::GenerateHueSaturationMatrix(double hue, double saturation, double coeff[3][3])
+{
+  double hue_shift                           = hue * pi_ / 180.0;
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:349]: pi_ is only used here, can we define it as local varible?

File: drm/DrmAtomicStateManager.cpp [hunk 24]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+{
+  double tmp[3][3];
+
+  for (int y = 0; y < 3; y++) {
+    for (int x = 0; x < 3; x++) {
+      tmp[y][x] = matrix_1[y][0] * matrix_2[0][x] + matrix_1[y][1] * matrix_2[1][x] + matrix_1[y][2] * matrix_2[2][x];
+    }
+  }
+
+  memcpy(result, tmp, sizeof(tmp));
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:344]: why it is needed?? can we remove tmp??

File: drm/DrmAtomicStateManager.cpp [hunk 25]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+  int read_bytes;
+  char buf[4096] = {};
+  double hue = 0.0;
+  double saturation = 100;
+  double coeff[3][3] = { { 1.0, 0.0, 0.0 }, { 0.0, 1.0, 0.0 }, { 0.0, 0.0, 1.0 } };
+
+  struct drm_color_ctm *ctm =
+    (struct drm_color_ctm *)malloc(sizeof(struct drm_color_ctm));
+  if (!ctm) {
+    ALOGD("Cannot allocate ctm memory");
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:380]: should be ALOGE

File: drm/DrmAtomicStateManager.cpp [hunk 26]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+  }
+
+  ALOGD("COLOR_ hue=%f", hue);
+  ALOGD("COLOR_ saturation=%f", saturation);
+
+  GenerateHueSaturationMatrix(hue, saturation, coeff);
+
+  for (int i = 0; i < 3; i++) {
+    for (int j = 0; j < 3; j++) {
+      color_transform_matrix[i * 4 + j] = coeff[j] [i];
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:426]: [j] and [i] add space??

File: drm/DrmAtomicStateManager.cpp [hunk 27]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+  if (contrast_c == 0 && brightness_c == 0) {
+    lut = NULL;
+    ApplyPendingLUT(lut, lut_size);
+    free(lut);
+    return 0;
+  }
+
+  lut = (struct drm_color_lut *)malloc(sizeof(struct drm_color_lut) * lut_size);
+  if (!lut) {
+    ALOGD("Cannot allocate LUT memory");
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:538]: should be ALOGE

File: drm/DrmAtomicStateManager.cpp [hunk 28]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+    return -ENOMEM;
+  }
+
+  /* Unpack brightness values for each channel */
+  temp[0] = (brightness_c >> 16) & 0xFF;
+  temp[1] = (brightness_c >> 8) & 0xFF;
+  temp[2] = (brightness_c)& 0xFF;
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:545]: coding style

File: drm/DrmAtomicStateManager.cpp [hunk 29]
```diff
@@ -326,4 +331,352 @@ auto DrmAtomicStateManager::ActivateDisplayUsingDPMS() -> int {
...
+
+  /* Map brightness from -128 - 127 range into -0.5 - 0.5 range */
+  brightness[0] = (float)(temp[0]) / 255 - 0.5;
+  brightness[1] = (float)(temp[1]) / 255 - 0.5;
+  brightness[2] = (float)(temp[2]) / 255 - 0.5;
+
+  /* Unpack contrast values for each channel */
+  temp[0] = (contrast_c >> 16) & 0xFF;
+  temp[1] = (contrast_c >> 8) & 0xFF;
+  temp[2] = (contrast_c)&0xFF;
```
Comment (feijiang1) [inline drm/DrmAtomicStateManager.cpp:555]: code style

File: drm/DrmDisplayPipeline.cpp [hunk 30]
```diff
@@ -203,7 +203,10 @@ auto DrmDisplayPipeline::AtomicDisablePipeline() -> int {
 
   if (!connector->Get()->GetCrtcIdProperty().AtomicSet(*pset, 0) ||
       !crtc->Get()->GetActiveProperty().AtomicSet(*pset,  0)||
-      !crtc->Get()->GetModeProperty().AtomicSet(*pset, 0)){
+      !crtc->Get()->GetModeProperty().AtomicSet(*pset, 0)||
+      !crtc->Get()->GetCtmProperty().AtomicSet(*pset, 0)||
+      !crtc->Get()->GetGammaLutProperty().AtomicSet(*pset, 0)||
+      !crtc->Get()->GetGammaLutSizeProperty().AtomicSet(*pset, 0)){
```
Comment (feijiang1) [inline drm/DrmDisplayPipeline.cpp:209]: before {, need space

**Review by feijiang1 — COMMENTED** [review]
some more comments

**Review by feijiang1 — APPROVED** [review]
LGTM

**Review by feijiang1 — APPROVED** [review]
LGTM

**Review by YuanjunHuang — APPROVED** [review]

**Review by feijiang1 — APPROVED** [review]
LGTM

**Review by feijiang1 — APPROVED** [review]
LGTM

**Review by feijiang1 — APPROVED** [review]
LGTM

**Review by iViggyPrabhu — APPROVED** [review]
Looks Good.

**Pull request merged**